scio-bigquery/src/it/scala/com/spotify/scio/bigquery/BigQueryIT.scala

```diff
@@ -0,0 +1,147 @@
+    shakespeare("thou", 20, "kinglear", 1600),
+    shakespeare("thy", 30, "kinglear", 1600))
+  val expected = Seq(wordCount("i", 10), wordCount("thou", 20), wordCount("thy", 30))
+
+  mockTable(tableRef).withData(inData)
+
+  "BigQueryITUtils" should "support mock data" in {
+    query(legacyQuery) should contain theSameElementsAs expected
+    query(sqlQuery) should contain theSameElementsAs expected
+  }
+
+}
+
+/**
+ * BigQuery integration test with sample data and property check for query result.
+ */
+class BigQuerySampleIT extends FlatSpec with Matchers with BigQueryITUtils {
+
+  import BigQueryIT._
+
+  mockTable(tableRef).withSample(100)
+
+  "BigQueryITUtils" should "support sample data" in {
+    forAll(query(sqlQuery)) { r =>
+      val word = r.get("word").toString
+      word should not be null
+      word should not be empty
+      r.get("word_count").toString.toInt should be > 0
+    }
+  }
+
+}
+
+
+/**
+ * Type-safe BigQuery integration test with mock data and equality check for query result.
+ */
+class BigQueryTypedIT extends FlatSpec with Matchers with BigQueryITUtils {
+
+  import BigQueryIT._
+
+  val inData = Seq(
+    Shakespeare("i", 10, "kinglear", 1600),
+    Shakespeare("thou", 20, "kinglear", 1600),
+    Shakespeare("thy", 30, "kinglear", 1600))
+  val expected = Seq(
+    WordCount(Some("i"), Some(10)),
+    WordCount(Some("thou"), Some(20)),
+    WordCount(Some("thy"), Some(30)))
+
+  mockTable(tableRef).withTypedData(inData)
+
+  "BigQueryITUtils" should "support typed BigQuery" in {
+    typedQuery[WordCount](legacyQuery) should contain theSameElementsAs expected
+    typedQuery[WordCount](sqlQuery) should contain theSameElementsAs expected
+  }
+
+}
+
+class BigQueryITUtilsTest extends FlatSpec with Matchers {
+
+  import BigQueryIT._
+
+  "BigQueryITUtils" should "fail insufficient sample data" in {
+    val t = "clouddataflow-readonly:samples.weather_stations"
+    the[IllegalArgumentException] thrownBy {
+      val it = new BigQueryITUtils {}
+      it.mockTable(t).withSample(2000)
+    } should have message s"requirement failed: Sample size 1000 != requested 2000"
+    the[IllegalArgumentException] thrownBy {
+      val it = new BigQueryITUtils {}
+      it.mockTable(t).withSample(2000, 5000)
+    } should have message s"requirement failed: Sample size 1000 < requested minimal 2000"
+  }
+
+  it should "fail duplicate mockTable" in {
+    val it = new BigQueryITUtils {}
+    it.mockTable(tableRef)
+    the[IllegalArgumentException] thrownBy {
+      it.mockTable(tableRef)
+    } should have message s"requirement failed: Table $tableRef already registered for mocking"
+  }
+
+  it should "fail duplicate mock data" in {
+    val it = new BigQueryITUtils {}
+    val mt = it.mockTable(tableRef)
+    mt.withData(Nil)
+    the[IllegalArgumentException] thrownBy {
+      mt.withData(Nil)
+    } should have message s"requirement failed: Table $tableRef already populated with mock data"
+  }
+
```

Comment:
If we register a mock table but don't supply any data, what's the expected behavior? Treat it as empty table and still allow querying? Should we include a test for this?